packages/zoe/src/contracts/barterExchange.js

```diff
@@ -0,0 +1,188 @@
+    }
+
+    function findMatchingTrade(newDetails, orders) {
+      return orders.find(order => {
+        const { amountIn, amountOut } = order;
+        // see if canTradeWith() would work as well
+        const amountMathNewIn = brandToAmountMath.get(newDetails.brandIn);
+        const amountMathNewOut = brandToAmountMath.get(newDetails.brandOut);
+        return (
+          amountMathNewIn.isGTE(newDetails.amountIn, amountOut) &&
+          amountMathNewOut.isGTE(amountIn, newDetails.amountOut)
+        );
+      });
+    }
+
+    function removeFromOrders(offerDetails) {
+      const orders = lookupBookOrders(
+        offerDetails.brandIn,
+        offerDetails.brandOut,
+      );
+      orders.splice(orders.indexOf(offerDetails), 1);
+    }
+
+    function crossMatchAmounts(offerDetails, matchingTrade) {
+      const amountMathNewIn = brandToAmountMath.get(offerDetails.brandIn);
+      const amountMathNewOut = brandToAmountMath.get(offerDetails.brandOut);
+      const newOfferAmountsRecord = {
+        [offerDetails.keywordOut]: offerDetails.amountOut,
+        [offerDetails.keywordIn]: amountMathNewIn.subtract(
+          offerDetails.amountIn,
+          matchingTrade.amountOut,
+        ),
+      };
+      const newMatchingAmountsRecord = {
+        [matchingTrade.keywordOut]: matchingTrade.amountOut,
+        [matchingTrade.keywordIn]: amountMathNewOut.subtract(
+          matchingTrade.amountIn,
+          offerDetails.amountOut,
+        ),
+      };
+      return [newOfferAmountsRecord, newMatchingAmountsRecord];
+    }
+
+    function tradeWithMatchingOffer(offerDetails) {
+      const orders = lookupBookOrders(
+        offerDetails.brandOut,
+        offerDetails.brandIn,
+      );
+      const matchingTrade = findMatchingTrade(offerDetails, orders);
+      if (matchingTrade) {
+        // reallocate by switching the amount
+        const amounts = crossMatchAmounts(offerDetails, matchingTrade);
+        const handles = [offerDetails.offerHandle, matchingTrade.offerHandle];
+        const keywords = [
+          [offerDetails.keywordIn, offerDetails.keywordOut],
+          [matchingTrade.keywordIn, matchingTrade.keywordOut],
+        ];
+        zcf.reallocate(handles, amounts, keywords);
+        // swap(offerDetails.offerHandle, matchingTrade.offerHandle);
+        removeFromOrders(matchingTrade);
+        zcf.complete(handles);
+
+        return true;
+      }
+      return false;
+    }
+
+    function addToBook(offerDetails) {
+      const orders = lookupBookOrders(
+        offerDetails.brandIn,
+        offerDetails.brandOut,
+      );
+      orders.push(offerDetails);
+    }
+
+    function extractBrandToAmountMaths(offerHandle, issuerKeywordRecord) {
+      const amountMathMap = new Map();
+      Object.getOwnPropertyNames(issuerKeywordRecord).forEach(keyword => {
+        const amountMath = zcf.getAmountMaths([keyword])[keyword];
+        amountMathMap.set(issuerKeywordRecord[keyword].getBrand(), amountMath);
```

Comment:
calling issuer.getBrand() will fail in a vat setting because the issuer is likely remote. I think we instead want to get the issuerRecord from Zoe which is synchronously accessible. 

However, it is probably better to change the `zcf.getAmountMaths `function and its usages to use brands instead of keywords. Perhaps instead of a function returning an AmountMathKeywordRecord, it's the `getAmountMathForBrand` function that you used in Zoe

Reply:
yup.